---
title: 'PathwaySplice: An R package for adjusting bias in pathway analysis using exon and splicing junction centric RNA-Seq data'
bibliography: Splicing-network.bib
header-includes: \usepackage{float}
output:
  html_document: default
  pdf_document:
    fig_caption: yes
    keep_tex: yes
  word_document: default
pandoc_args: --natbib 
biblio-style: plain
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.pos = 'H')
```

## Motivation: 
Differential exons or splicing junctions usage analysis is very helpful for detecting alternative splicing events. There are several studies that researchers attempt to undertand alternative splicing events by performing differential exons or/and splicing junctions usage analysis. In these analysis, the differentially expressed genes are derived based on differential exon or/and splicing junction usage analysis, then subsequent gene set or/and pathway analysis are performed based on differentlly expressed gene list. However, it is observed that the derived differentially expressed gene list from differential exon or/and splicing junction analysis is biased by the number of exons or/and splicing junctions within gene, and lead to further bias in subsequent get set and pathway analysis.

![Bias](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/bias.png)

## Results: 
We demonstrate this bias using a published data from a study of differential splicing junction usage in Myelodysplastic syndromes(MDS). We show that the identified differentially expressed genes through using differential usage of gene features(exons and/or splicing junctions) are biased by the number of exons and splicing junctions within gene, and subsequently lead to the bias in Gene Ontology(GO) and Kyoto Encyclopedia of Genes and Genomes(KEGG) enrichment analysis. We develop an approach to adjust this bias. An R package(PathwaySplice) based on this approach is also implemented to make this method be available for other researchers. This PathwaySJ package is accessable on <https://github.com/aiminy/PathwaySplice.git>.

## Introduction

Alternative Splicing is an important post-transcriptional regulation process. This process allows one gene to produce mutiple mRMA transcripts, and further contributes to protein diversity. The previous researches have showed more than 95% of human genes are alternatively spliced. RNA-Seq data has demonstrated great advantages to study more complex regulatry phenomemnea at transcript isoforms, exon and splicing junction level besides studying gene-wide effects. Numerous tools have been developed to use RNA-seq data to perform alternative splicing analysis. These tools can be classified into two categories. The first approach focuses on transcript isoform expression[@Zheng2009]. The second method skips estimaing transcript isoforms expression, and explore to detect signals of alternative splicing using the difference on the distrbutions of reads on exons and junctions of genes bwteen conditions[@Anders2012][@Hartley2016]. As an important remedy for the analysis on gene and transcript level, exon-centric or /and splicing junction-centric approaches can supply more information than gene and transcript level, and is very valuable for biological interpretation. 

In the trancriptomics studies on gene-level,a standdard step is to understand the molecular functions or pathways in which identified differentially expressed genes(DEGs) are involved. Since genes usually work together in biological process, so functional profiling analysis using identified differential genes becomes a critical component for biological interpretation of DEGs. There are several previous studies that showed using RNA-Seq data to study differentially expressed genes and subsequent gene sets or pathways analysis could be biased by read counts of transcripts and length of transcripts. Several methods are also developed to correct these biases either in the step of identifing differentially genes or in the step of performing gene sets enrichment analysis[@Gao2011]. A well known R package for correcting bias in gene set enrichment analysis, GoSeq, is available[@Young2010]. A logistic regression based approach is also developed for adjusting length bias of transcirpt in Gene Ontology enrichment analysis[@Mi2012].

Likewisely, several researchers extend the idea for functional profiling in gene-based approach to the exon-centric or/and splicing junction-centric approaches. In these studies, they identified differentially used exons and splicing junctions with gene, and derived DEGs based on differentll usage of exons and junctions, and performed the functional profiling analysis using these DEGs. In their analysis, they taks it for granted that gene length is a bias factor as in gene-level based analysis. However, the way to identify DEGs based differential usage of exons and junctions could make gene length be an unsuitable bias factor.

Here we use a real data set to address the existing problems in pathway analysis from exon and splicing junction based DEGs, and propose an analysis approach to identify the suitable bias factors. We also implementate this procedure into an R package to let this approach be available to other researchers.

## RESULTS AND DISCUSSIONS

## Analysis on a real data set


We use a published data set to demonstrate our approach. This data set is from studying key genes and pathways in myelodysplastic syndrome hermatopoietic stem and progenenitor cells. The data set has 16 human subjects. Among them, there are 5 health controls, 4 SF3B1 wild types and 7 SF3B1 mutants. The scientists attempt to understand the effect of SF3B1 mutation on alternative splicing. 
To achieve this goal, we performed analysis to uncover differential usage of exons and/or splicing junction for the comparisions between SF3B1 mutants with SF3B1 wild type and SF3B1 mutants with health controls by using DEXSeq R package. After this step, we use the p values for each exon and/or splicing junction with each gene to represent the p value of gene either using smallest p value or using genewise FDR caculated by using smallest and number of exon and/or splicing junctions. However, we observe that differential genes using this procedure is biased by the number of exons and/or splicing junctions. We use the comparision between SF3B1 mutation with widetype to show this bias:

![Bias_from_number_of exons](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/Bias_gene.png)

The above Figure shows differential genes derived using exon splicing junction related p values is biased by the number of exons within gene. The larger number of exons within gene is, the smaller the p value of exon is. 

### Adjust bias on the step for identifying DGE

Since the distribution of p value is uniform bounded bwtween 0 and 1, and the distribution of number of exons follows a possion distribution, so it is not trival to identify the relationship between p value and number of exons. Therefore we attempt to uncover the relationship between test statistics and number of exons. The following shows the relationship betweeen test statistics and number of exons.

![Relationship_between_statistics_and_exons](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/statistics_exon.png)

Pearson correlation between test statistcs and number of exons is 0.43, which indicates these exists certain relationship between test statistics and number of exons. Here we use Y to indicate test statistics and use X to indicate number of exons and/or other possible bias factors. We attempt to fit the follwoing model to maximizxe the proportion of variance of Y that is explained by varince of X.

$$Y = \beta_0+\sum\limits_{i=1}^{p}\beta_{i}X_{i}+\epsilon$$

In the situation for considering number of exons only, $p$ will be 1. After this fitting, the effect of X on Y will be regressed out, so the adjusted test statistics can obtained by the following:

$$\tilde{Y_{i}}=\bar{Y}+\tilde{\epsilon_{i}}$$

Here $\bar{Y}$ is the grand mean of Y, and $\tilde{\epsilon_{i}}$ is the residual from the above fitted model.

Firstly, we fit Y to X based on a linear regression model, then we can get adjusted test statistics, p value can be calculated using this adjusted statistics. Using these adjusted p-values, we identidy differential genes. Afterwards, we use logistic regression to check whether the effect from number of exons on differential gene list still exists.

![adjust_based_linear](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/linear.png)

It is obvious that fitting a linear regression model does not completely regress out the effect of number of exons. This result also indicates that fitting a linear regression model does not maximizxe the proportion of variance of Y that is explained by varince of X. So we need more suitable methods to achieve this goal.
From the scatter plot for test statistics and number of exons, we can see that the possible nonlinear relationship could be existed, so we fit a polynomial regression model

![cubic](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/cubic.png)

In this model, we use 2.99 as the degree of this polynomial regression model, we can see some adjustment fot this setting. However certain trend still can be observed. So we explore to apply Alternating Conditional Expectations(ACE) regresssion model to this problem. The following is ACE model:

$$\theta(Y) = \alpha+\sum\limits_{i=1}^{p}\phi_i(X_{i})+\epsilon$$

$$\epsilon^2(\theta,\phi_{1},...,\phi_{p})=E\{[\theta(Y)-\sum\limits_{i=1}^{p}\phi_i(X_{i})]\}^2$$

$$\phi_{i}(X_{i})=E[\theta(Y)-\sum\limits_{{j} \ne {i}}^{p}\phi_j(X_{j})|X_{i}]$$

$$\theta(Y)=E[\sum\limits_{i=1}^{p}\phi_i(X_{i})|Y]/||E[\sum\limits_{i=1}^{p}\phi_i(X_{i})|Y]||$$

Two above operations are conditional expectations and iterative minimisation and hence, the name Alternating Comnditional Expectations(ACE). Finally, we get optimal transformation for $\phi_{i}^*(X_{i}),i=1,...,p,$ and $\theta^*(Y)$. In the transformed space, Y and X are related as follows:

$$\theta^*(Y)=\sum\limits_{i=1}^{p}\phi_i^*(X_{i})+\epsilon^*$$

Here ${\epsilon^*}^2=1-{\rho^*}^2$. We use ACE method to get adjusted test statistics, and perform  similar procedures as above to check whether the effect from number of exons on differential gene list are regressed out. 

![ACE](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/ACE.png)

Pearson correlation between ACE transformed test statistcs and number of exons is 0.6142665, which is greater than the Pearson correlation(0.43) based on untransformed test statistcs and number of exons,which indicates ACE model captures the complicated relatioship between test statistcs and number of exons. 

![Adjust_Bias_on_DE](/Volumes/Bioinformatics\$/Aimin_project/GOSJ_example_data/presentation/P_adjused.png)

It is clear the effect of number of exons on DGE analysis are regressed out.

### Adjust bias on the step for perforing gene set enrichment analysis 

All adjustment we performed here is based on adjusting bias on the different gene list. One common task after DGE identification is to use gene list to perform pathway and GO term analysis. Since DGE is already biased, so subsequent pathway is biased also.

Here we use these differnentially expressed gene list to identify the relationship between the proportion of DE in gene sets and the possible bias factos such as gene length, number of exons and number splicing junctions. 

![DESJ](/Volumes/Bioinformatics$/Aimin_project/GOSJ_example_data/Results/Pro_DE-vs-SJ.png)

<!-- ![FeatureGL](/home/aiminyan/GOSJ/Figure/pwfFeatureGL.tiff) -->

<!-- ![FeatureFeature](/home/aiminyan/GOSJ/Figure/pwFeatureFeature.tiff) -->


<!-- ![Bias from sj and GL](/home/aiminyan/GOSJ/Figure/sjvsGL.png) -->

<!-- ![Bias from sj and exons](/home/aiminyan/GOSJ/Figure/sjvsexons.png) -->

It is obvious that DEGs is biased by the number of exons and spling junctions with gene. We use this information adjust subsequent gene set and pathway analysis. In this package, we implement several functions to make this package be able to perform differnt gene set and pathway definitions. 

+ Correction for GO term enrichment analysis

We implemented a function to perform GO term enrichment analysis, and adjust this analysis by the number of splicing junctions. Using this function, user can obtain the information about enriched GO, the number of genes and the number of diffenerially expressed genes of each GO.

![EnrichedGO](/Volumes/Bioinformatics$/Aimin_project/GOSJ_example_data/Results/EnrichED._GO.png)

+ Correction for KEGG pawhay anlaysis.

We also implemented a function to let user be able to perform KEGG pathway analysis by adjusting the number of splicing junctions of each gene.

+ Correction for hallmark gene sets

Besides GO and KEGG pathway analysis, we implemented a function to let user be able to choose other gene sets to perform the number of splicing junction-based gene set adjustment analysis.

+ Enrichmap build-up and visualization

GOseq package does not supply the functionalities to construct and visualize network based on the overlap on genes between the enriched gene sets calculated using suitable bias factor. 
Moreover, GOseq does not inlcude differential genes within each gene set in its output. For building and visualizing functional network from enriched gene sets, Yu developed the clusterProfiler R package to compare and visualize functional categories of gene sets (Yu, et al., 2012). In clusterProfiler, enrichment test for GO and KEGG pathways is based on hypergeometric distribution.  For the enrichment analysis of GO term and pathway based on the differential gene list derived from differential exon and/or splicing junctions, enrichment analysis is performed by taking the effect of bias factors such as number of exons and/or splicing junctions into consideration. clusterProfier does not supply streamlined procedure to build up network for enrichment analysis from adjusting these bias factors. Therefore, these procedures need to be implemented for unbiased gene set or pathway analysis.       
 
In PathwaySplice, we fill in these gaps to let user be able to perform these analysis.

![Enrichedmap_adjusted](/Users/axy148/PathwaySplice/Network_adjusted.png)

![Enrichedmap_unadjusted](/Users/axy148/PathwaySplice/Network_unadjusted.png)

![Enrichedmap_cp_adjusted](/Users/axy148/PathwaySplice/cp_pathway_adjusted.png)

![Enrichedmap_cp_unadjusted](/Users/axy148/PathwaySplice/cp_pathway_unadjusted.png)

<!-- ## Generate permutated data set based on the data set from blood cancer.  -->

<!-- Using the idea in [@Geeleher2013], we explore to fix gene structure in the data set in blood cancer, and use permutation to generate more data sets to demostrate the bias from gene structure. We have about 13000 genes, each gene has certain number of subfeatures, in this permutation setting, we fix the number of subfeatures for each gene as the real data, and perform the following permutations: -->

<!-- + Permutating sample label -->

<!-- + Permutating mapping counts  -->

<!-- By this way, we explore to identify the same enriched gene sets and pathways of GO and KEGG. If there are high consistencies between the resutls from permutated data sets and that of real data sets, this indicates that the number of subfeatures of genes biases gene set or/and pathway analysis. -->

## Conclusion
In summary, our work identified the potential bias in RNA-Seq data analysis when using differential exons or/and splicing junction usage to represent differential expressed gene, and using these differential expressed genes to perform subsequent gene set and pathway analysis. We futher sugguest a bias correction approach that could provide a more accutate gene set and pathway analsis. We implemented this method into an R package, and we believe that this R packcage could help other researchers when they perform similar type of RNA-Seq data analysis.

## Reference
