---
title: "PathwaySplice"
author: "Aimin Yan"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PathwaySplice}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(fig.width=12,fig.height=8,collapse = FALSE, comment = "#>")
```

# Use PathwaySplice

+ Convert feature based data to gene based data
```{r eval=TRUE,warning=FALSE,message=FALSE,results='hide'}
library(PathwaySplice)
data("featureBasedData")
gene.based.table <- makeGeneTable(featureBasedData)
```

+ Apply logistic regression model to identify bias factor
```{r eval=TRUE,warning=FALSE,message=FALSE,results='hide'}
lrTestBias(gene.based.table,boxplot.width=0.3)
```

+ Perform analysis using pathwaysplice
```{r eval=TRUE,warning=FALSE,message=FALSE,results='hide'}

# This function performes gene set analysis that adjusts for different number of gene features 
# (i.e. exons and/or spicing junctions) associated with each gene. 
# The significant genes are those with smallest p-value of gene feature  < *threshold*.

res <- runPathwaySplice(gene.based.table,genome='hg19',id='ensGene',test.cats=c('GO:BP'),go.size.limit=c(5,30),method='Wallenius')
```

+ Build up network based on the overlap between gene sets and visualize this network

```{r eval=TRUE,warning=FALSE,message=FALSE,results='hide'}
#test
output.file.dir <- file.path(tempdir(),"OutputEnmap")
enmap <- enrichmentMap(res,n=3,output.file.dir
                       =output.file.dir,similarity.threshold=0)
```